File: 0647-palindromic-substrings/0647-palindromic-substrings.swift
class Solution {
    func countSubstrings(_ s: String) -> Int {
        let n = s.count
        var count = 0
        for i in 0..<n {
            count += countPalindromes(s, i, i)
            count += countPalindromes(s, i, i + 1)
        }
        return count
    }

    private func countPalindromes(_ s: String, _ left: Int, _ right: Int) -> Int {
        var count = 0
        var left = left
        var right = right
        while left >= 0 && right < s.count && s[s.index(s.startIndex, offsetBy: left)] == s[s.index(s.startIndex, offsetBy: right)] {
            count += 1
            left -= 1
            right += 1
        }
        return count
    }
}

File: 0647-palindromic-substrings/0647-palindromic-substrings.dart
class Solution {
  int countSubstrings(String s) {
    final int n = s.length;
    int count = 0;
    for (int i = 0; i < n; i++) {
      count += _countPalindromes(s, i, i);
      count += _countPalindromes(s, i, i + 1);
    }
    return count;
  }

  int _countPalindromes(String s, int left, int right) {
    int count = 0;
    while (_isValid(left, right, s)) {
      left--;
      right++;
      count++;
    }
    return count;
  }

  bool _isValid(int left, int right, String s) =>
      left >= 0 && right < s.length && s[left] == s[right];
}
